File: CompareLang/Animal.java
abstract class Animal {

    public Animal() { //constructor
        System.out.println("Animal constructor called.");
    }
    @Override //destructor
    public void finalize() throws Throwable {
        System.out.println("Animal destructor called.");
    }
    public abstract void makeSound();//pure virtual function
    public void eat() {//virtual function
        System.out.println("This animal eats food.");
    }
    final void breathe() {//non virtual function
        System.out.println("(Breathing...)");
    }

}

class Dog extends Animal {
    public Dog() {
        System.out.println(">> Dog constructor called.");
    }
    @Override
    public void finalize() throws Throwable {
        System.out.println("<< Dog destructor called.");
    }
    @Override
    public void makeSound() {
        System.out.println("Woof!");
    }
    @Override
    public void eat() {
        System.out.println("The dog eats dog food.");
    }
}
class Cat extends Animal {
    public Cat() {
        System.out.println(">> Cat constructor called.");
    }

    @Override
    public void finalize() throws Throwable {
        System.out.println("<< Cat destructor called.");
    }

    @Override
    public void makeSound() {
        System.out.println("Meow!");
    }
}
class Labrador extends Dog {
    public Labrador() {
        System.out.println(">>> Labrador constructor called.");
    }

    @Override
    public void finalize() throws Throwable {
        System.out.println("<<< Labrador destructor called.");
    }
    @Override
    public void makeSound() {
        System.out.println("Labrador Woof!");
    }
}

class Main {
    public static void main(String[] args) {
        //Java has only heap(1-3)
        Dog dog = new Dog();
        System.out.println("dog(object) is created.");

        Labrador labrador = new Labrador();
        System.out.println("labrador(object) is created.");

        Animal cat = new Cat();
        System.out.println("cat(object) is created.");

        Animal a[] = new Animal[3];
        a[0]=dog;
        a[1]=labrador;
        a[2]=cat;

        for(int i=0;i<a.length;i++){
            System.out.println("--- Animal number ["+i+"] ---");
            a[i].makeSound();
            a[i].eat();
            a[i].breathe();
        }
        
        dog = null;
        labrador = null;
        cat = null;

        System.gc();//ขอให้ garbage collector ทำงาน
        //ก่อนถูกลบ JVM อาจเรียก เมธอด finalize() ของ object นั้น
        System.out.println("Heap memory deleted successfully.");
    }
}

File: CompareLang/Animal.cpp
#include <iostream>
using namespace std;

// Abstract Class
class Animal {
public:
    Animal() { // constructor
        cout << "Animal constructor called." << endl;
    }
    virtual ~Animal() { // destructor
        cout << "Animal destructor called." << endl;
    }
    virtual void makeSound() = 0; // pure virtual function (override ได้)
    virtual void eat() {// virtual function (override ได้)
        cout << "This animal eats food." << endl;
    }
    void breathe() {// non-virtual function (ห้าม override)
        cout << "(Breathing...)" << endl;
    }
};

class Dog : public Animal {
public:
    Dog() {
        cout << ">> Dog constructor called." << endl;
    }

    ~Dog() override {
        cout << "<< Dog destructor called." << endl;
    }

    void makeSound() override {
        cout << "Woof!" << endl;
    }

    void eat() override {
        cout << "The dog eats dog food." << endl;
    }
};

class Cat : public Animal {
public:
    Cat() {
        cout << ">> Cat constructor called." << endl;
    }

    ~Cat() override {
        cout << "<< Cat destructor called." << endl;
    }

    void makeSound() override {
        cout << "Meow!" << endl;
    }
};

class Labrador : public Dog {
public:
    Labrador() {
        cout << ">>> Labrador constructor called." << endl;
    }

    ~Labrador() override {
        cout << "<<< Labrador destructor called." << endl;
    }

    void makeSound() override {
        cout << "Labrador Woof!" << endl;
    }
};

int main() {
    // Static allocation
    static Dog dog;
    cout << "dog(object) is created." << endl;

    // Stack allocation,Destructor จะถูกเรียกอัตโนมัติเมื่อออกจากฟังก์ชัน
    Labrador labrador;
    cout << "labrador(object) is created." << endl;

    // Heap allocation 
    Animal* cat = new Cat();
    cout << "cat(object) is created." << endl;

    // Polymorphism ผ่าน base class p
    Animal* a[3];
    a[0] = &dog;   // static
    a[1] = &labrador; // stack
    a[2] = cat;  //heap
    for (int i = 0; i < 3; ++i) {
        cout << "--- Animal number [" << i << "] ---" << endl;
        a[i]->makeSound();
        a[i]->eat();
        a[i]->breathe();
        cout << endl;
    }

    cout << "\n=== Memory Cleanup ===" << endl;
    delete cat;
    cout << "Heap memory deleted successfully." << endl;
    return 0;
}
